### PR TITLE
Pairtools module

### PR DESCRIPTION
**Pull request by sergpolly**
Hey @ewels !
this is a "Work in progress" for the [pairtools](https://github.com/mirnylab/pairtools) module:
 - `pairtools` is a set of command line framework for common operations on ["pairs"](https://github.com/4dn-dcic/pairix/blob/master/pairs_format_specification.md) generated by Hi-C and other "Chromatin Conformation Capture" ("C") techniques.
 - "Pairs" can be viewed as an [intermediate](https://pairtools.readthedocs.io/en/latest/) result of processing raw Hi-C (and other "C") data: raw paired-end "C"-data reads mapped to reference genome are parsed into "pairs", which can be further classified, sorted, flipped, deduplicated and filtered - these operations are the main domain of `pairtools`.
 - `Pairtools` is also responsible for collecting summary-statistics (as part of pipelines [[1]](https://data.4dnucleome.org/resources/data-analysis/hi_c-processing-pipeline) [[2]](https://github.com/mirnylab/distiller-nf), or [standalone](https://pairtools.readthedocs.io/en/latest/cli_tools.html#pairtools-stats)) and MultiQC is an amazing platform to aggregate, visualize and combine such pair-level stats with others (fastqc, adapter trimming, etc).

`Pairtools` is written by @golobor and influenced by many others from [4DN](http://dcic.4dnucleome.org/), [mirnylab](https://github.com/mirnylab/), etc

PS. I'm hoping that having this as a WIP PR would motivate/speed up the development of this module

Thank you so much!

## If this PR is for a new module
 - [x] There is example tool output for tools in the https://github.com/ewels/MultiQC_TestData repository
 - [x] Code is tested and works locally (including with `--lint` flag)
 - [x] `CHANGELOG.md` is updated
 - [x] `README.md` is updated
 - [x] `docs/README.md` is updated with link to below
 - [x] `docs/modulename.md` is created
 - [x] Everything that can be represented with a plot instead of a table is a plot
 - [x] Report sections have a description and help text (with `self.add_section`)
 - [x] There aren't any huge tables with > 6 columns (explain reasoning if so)
 - [x] Each table column has a different colour scale to its neighbour, which relates to the data (eg. if high numbers are bad, they're red)
 - [x] Module does not do any significant computational work

### CODE REVIEW

**Comment by ewels** [comment]
Hi @sergpolly,

Thanks for this - I had a look at your example report in another issue the other day and it looked beautiful! 🤩 

> PS. I'm hoping that having this as a WIP PR would motivate/speed up the development of this module

Is this PR still a WIP or is it ready for review?

Phil

**Comment by sergpolly** [comment]
Hi!
sorry for delay...
yes, this quarantine-edition PR is still WIP - I have to clean up the code (a lot) and incorporate some suggestions from the initial feedback (BTW - people who experienced MultiQC for the first time loved it - experimentalists and computational people alike)

@ewels do you have any approximate timeline for `1.9` release or maybe some window of time when you'll be focusing on `multiqc` ? I'm just trying to come up with some deadline for myself))

**Comment by ewels** [comment]
Ok great. Glad to hear that it’s proving popular!

As of yesterday a MultiQC is now at the top of my list, so I’m afraid that the timeframe is rather short now. I’m hoping to finish up for a release next week if I can, or maybe the week after.

Phil

**Comment by remiolsen** [comment]
Hi @sergpolly and @ewels 

I'm really interested in having a pairtools multiqc module. What is the status of this PR?

/Remi

**Comment by sergpolly** [comment]
Hi @remiolsen - I'm getting back to it after a long break - will try to submit it for review before 2022
the module itself was working all this time (installable from our branch), I just couldn't finalize it

File: multiqc/modules/pairtools/pairtools.py [hunk 1]
```diff
@@ -0,0 +1,522 @@
...
+            # fill in the data for XY-line plotting
+            # i.e. dict by samples of dicts by dist (X), of (normalized) counts (Y):
+            for cat in data_cats:
+                _data[cat][s_name] = dict(zip(_dist_bins_geom[1:], _summary[cat][2:]))
+
+                log.critical(_dist_bins_geom)
+
+        pconfig = {
+            'id': 'broom_plot',
+            'title': 'Pairs by distance and by read orientation',
```
Comment (ewels) [inline multiqc/modules/pairtools/pairtools.py:375]: Needs this plot title format to pass MultiQC `--lint` tests:

```suggestion
            'title': 'Pairtools: Pairs by distance and by read orientation',
```

File: multiqc/modules/pairtools/pairtools.py [hunk 2]
```diff
@@ -0,0 +1,522 @@
...
+            'xlab': 'Genomic separation (bp)',
```
Comment (ewels) [inline multiqc/modules/pairtools/pairtools.py:376]: Need a y-axis label to pass MultiQC `--lint`:

```suggestion
            'xlab': 'Genomic separation (bp)',
            'ylab': 'Number of reads',
```

**Review by ewels — COMMENTED** [review]
Hi @sergpolly,

Just checking in to see how you're doing with the PR. Let me know if there's anything I can do to help.

I just tried to fix the merge conflicts for you, but I don't have permissions to push to your repo as it's coming from an organisation fork rather than a personal one. It's a simple one though, just the changelog. Also make sure you use `-` for the markdown bullet there instead of `*` else the markdown linting will fail.

The other thing that you need to do is to run [black](https://black.readthedocs.io/en/stable/):

```console
$ pip install black
$ black .
reformatted multiqc/modules/pairtools/utils.py
reformatted multiqc/modules/pairtools/pairtools.py

All done! ✨ 🍰 ✨
2 files reformatted, 344 files left unchanged.
```

There are a couple of minor MultiQC lint errors too, see review comments.

Phil

**Comment by mikewolfe** [comment]
Hey @sergpolly and @ewels,

I recently started using `pairtools` as part of a Hi-C pipeline I am working and would love to have `multiQC` support for the stats.

I see that this module appears just about ready to go and was wondering if there was anyway it was going to be finished anytime soon? Anything I can do to help?

**Comment by ewels** [comment]
I confess I've been ignoring it as it's still marked as a Draft, I was basically waiting for @sergpolly to tell me it was ready for review + merge.

I agree that it looks basically good to go from my side though, so maybe we should go for it.

**Comment by mikewolfe** [comment]
To give my two cents: Cloning this repository and running this version of `multiqc` works pretty well using stats output from `pairtools` version 1.0.2. 

Great and informative quality control plots for Hi-C data.

**Comment by ewels** [comment]
Hi all,

I just moved to PR https://github.com/ewels/MultiQC/pull/1834 so that I could push edits directly (no permissions on this PR). I've got it up to date with the main repo and have fixed a bunch of minor bugs, but the more I'm digging into the code the more work I'm finding.

My main concern is that the module is doing too much work - there's a lot of stuff happening with numpy (which is causing me headaches as the data structures are not easily exported into JSON). But honestly, I feel like this code should be in the main tool. The concept behind MultiQC is just to scoop up stats and report on them, not to calculate them in the run time.

If @sergpolly could move most of this code out and keep the MultiQC module as lightweight as possible, it would be mean easier maintenance for him and many fewer problems with this module.

Phil

**Comment by sergpolly** [comment]
Hi !
I think this PR is ready for review again - we've simplified it and made it conform all of the linting/pre-commit business.
Apparently due to https://github.com/orgs/community/discussions/5634 - we cannot simply "allow edits by maintainers", so instead I just gave @ewels and @vladsavelyev write access to our fork
Corresponding test sample data has been merged already to test-data repo
Thank you so much for consideration and Happy 2024 !

**Comment by mikewolfe** [comment]
Hey @sergpolly and @ewels,

Any chance of this module making it into the next Multiqc release?

Thanks for all your hard work!

**Comment by vladsavelyev** [comment]
Sorry it's taking so long to merge this PR! I looked at it again, merged with main and slightly reformatted, fixed minor bugs. Should be good to merge now 👍

**Pull request merged**